daffodil-runtime1/src/main/scala/edu/illinois/ncsa/daffodil/processors/parsers/PackedBinaryTraits.scala

```diff
@@ -0,0 +1,190 @@
+import edu.illinois.ncsa.daffodil.exceptions.Assert
+import edu.illinois.ncsa.daffodil.equality._
+import edu.illinois.ncsa.daffodil.processors.dfa
+import edu.illinois.ncsa.daffodil.util.MaybeChar
+import edu.illinois.ncsa.daffodil.schema.annotation.props.gen.BinaryNumberCheckPolicy
+import java.lang.{ Number => JNumber }
+import java.math.{ BigInteger => JBigInteger, BigDecimal => JBigDecimal }
+import java.nio.charset.StandardCharsets
+import passera.unsigned.ULong
+
+trait PackedBinaryConversion {
+  def toBigInteger(num: Array[Byte], signCodes: String, policy: BinaryNumberCheckPolicy): JBigInteger
+  def toBigDecimal(num: Array[Byte], scale: Int, signCodes: String, policy: BinaryNumberCheckPolicy): JBigDecimal
+}
+
+abstract class PackedBinaryDecimalBaseParser(
+  e: ElementRuntimeData,
+  binaryDecimalVirtualPoint: Int,
+  binaryPackedSignCodes: String = "",
+  binaryNumberCheckPolicy: BinaryNumberCheckPolicy = BinaryNumberCheckPolicy.Strict)
+  extends PrimParserObject(e)
+  with PackedBinaryConversion {
+
+  protected def getBitLength(s: ParseOrUnparseState): Int
+
+  def parse(start: PState): Unit = {
+    val nBits = getBitLength(start)
+    if (nBits == 0) return // zero length is used for outputValueCalc often.
+    val dis = start.dataInputStream
+
+    if (!dis.isDefinedForLength(nBits)) {
+      PE(start, "Insufficient bits in data. Needed %d bit(s) but found only %d available.", nBits, dis.remainingBits.get)
+      return
+    }
+
+    val bigDec = toBigDecimal(dis.getByteArray(nBits, start), binaryDecimalVirtualPoint, binaryPackedSignCodes, binaryNumberCheckPolicy)
+    start.simpleElement.overwriteDataValue(bigDec)
+  }
+}
+
+abstract class PackedBinaryIntegerBaseParser(
+  e: ElementRuntimeData,
+  signed: Boolean = false,
+  binaryPackedSignCodes: String = "",
+  binaryNumberCheckPolicy: BinaryNumberCheckPolicy = BinaryNumberCheckPolicy.Strict)
+  extends PrimParserObject(e)
+  with PackedBinaryConversion {
+
+  protected def getBitLength(s: ParseOrUnparseState): Int
+
+  def parse(start: PState): Unit = {
+    val nBits = getBitLength(start)
+    if (nBits == 0) return // zero length is used for outputValueCalc often.
+    val dis = start.dataInputStream
+
+    if (!dis.isDefinedForLength(nBits)) {
+      PE(start, "Insufficient bits in data. Needed %d bit(s) but found only %d available.", nBits, dis.remainingBits.get)
+      return
+    }
+
+    val int: JNumber =
+      if (signed) {
+        if (nBits > 64) { toBigInteger(dis.getByteArray(nBits, start), binaryPackedSignCodes, binaryNumberCheckPolicy) }
+        else { toBigInteger(dis.getByteArray(nBits, start), binaryPackedSignCodes, binaryNumberCheckPolicy).longValue() }
+      } else {
+        if (nBits >= 64) { toBigInteger(dis.getByteArray(nBits, start), binaryPackedSignCodes, binaryNumberCheckPolicy) }
+        else { toBigInteger(dis.getByteArray(nBits, start), binaryPackedSignCodes, binaryNumberCheckPolicy).longValue() }
+      }
```

Comment:
I'm not sure if these sign checks make sense. This is just getting a byte array and converting to a BigInteger. Why the signed check and conversion to long?

Also, what happens if the toBigInteger returns a negative value but ``signed`` is false? Should that be an SDE?

Also, I'm not sure the nBits check makes sense.  Does >64 bits of data mean a packed value can't be converted to a long? Why even bother converting to a long? Does something else depend on that?

Reply:
Another thing related to this. I think we need to check to make sure that the parsed number is without the bounds of the infoset type. For example, assume the schema type is an xs:long. We parse nBits of data and successfully convert it to a number. It's possible that that number is out of bounds for an xs:long. We either need to determine the number of bits of these packed/zoned types that can fit in each infoset type and check at schema compilation type, or if that isn't possible, after converting to an BigInt/Decimal we need to validate that it is within range of the infoset type at runtime.

Reply:
Also related, the table in Section 13.7  bcd is not valid for xs:long, xs:int, xs:short, and xs:byte, so those should be an SDE. Note that xs:decimal and xs:integer are allowed for BCD, so we still need to do checks on unparse if the infoset value is negative.